types/wlr_input_method_v2.c

```diff
@@ -101,10 +122,196 @@ static void im_get_input_popup_surface(struct wl_client *client,
 	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::get_input_popup_surface");
 }
 
+static void keyboard_grab_destroy(
+		struct wlr_input_method_keyboard_grab_v2 *keyboard_grab) {
+	wlr_signal_emit_safe(&keyboard_grab->events.destroy, keyboard_grab);
+	keyboard_grab->input_method->keyboard_grab = NULL;
+	if (keyboard_grab->keyboard) {
+		wl_list_remove(&keyboard_grab->keymap_listener.link);
+		wl_list_remove(&keyboard_grab->repeat_info_listener.link);
+
+		// send modifier state to original client
+		wlr_seat_keyboard_notify_modifiers(keyboard_grab->input_method->seat,
+			&keyboard_grab->keyboard->modifiers);
```

Comment:
Hmm, this does things behind the compositor's back. I'd rather not.

Reply:
I've moved this to compositor side.